

<left>FINM 36700 - Portfolio Theory and Risk Management</left> 
<br>
<left>Fall 2023</left>

<h2><center> Homework 8 - Long-Term Capital Management, L.P. </center></h2>



<h3><span style="color:#00008B">GroupC 26</span></h3>


### 1.1) Describe LTCM's investment strategy with regard to the following aspects:
- #### Securities traded

LTCM(Long term capital management) focused on exploiting market inefficiencies and engaging in arbitrage through Relative Value and Convergence trades. They adopted a long-short trading strategy to capitalize on these slight mispricings, employing significant leverage to amplify their investments in these minor discrepancies. Their strategy also involved hedging their positions through these long-short trades.

Moreover, LTCM was deeply engaged in income and credit markets, along with substantial involvement in equities. They traded a wide array of securities across various global markets.

- #### Trading frequency

LTCM's trading frequency was closely aligned with their distinct strategies. Their primary investment approach, convergence trades, typically had a long-term trading horizon, spanning weeks or months. Generally, LTCM did not focus on arbitraging intraday market movements, nor did they engage in long-term directional betting.

- #### Skewness (Do they seek many small wins or a few big hits?)

LTCM's strategy was oriented towards achieving small, positive returns by employing leverage, rather than placing significant bets on specific events. This approach resulted in a lower skewness in their returns compared to instruments like the S&P 500 Index (SPY). However, this strategy made them more vulnerable to extreme market events.

- #### Forecasting (What is behind their selection of trades?)


LTCM's approach to selecting trades was rooted in building models to identify mispricing in the market and understanding the reasons behind such mispricing. They would then forecast their potential profit and loss based on these identified opportunities. The key to their forecasting wasn't solely their advanced mathematical models; while they utilized concepts like convergence trade and relative value theory, their real edge came from their deep knowledge and understanding of market dynamics.

### 1.2) What are LTCM's biggest advantages over its competitors?


LTCM's primary advantages over its competitors included:

Efficient Financing: LTCM excelled in securing financing and funding, complemented by their proprietary trading and modeling capabilities.

Fund Size: Their large Assets Under Management (AUM) enabled them to leverage at favorable rates, providing a significant advantage in scale and negotiating power.

Collateralization: LTCM was adept at better collateralizing their positions, resulting in lower haircuts and improved credit terms.

Long-term Horizon: They had a long-term commitment of capital from investors, coupled with the availability of extended credit lines, allowing for more strategic and less pressured investment decisions.

Liquidity and Hedging: LTCM implemented various mechanisms to ensure liquidity in their operations. They also focused on minimizing default risk and refrained from making explicit directional bets, enhancing their risk management and stability.

### 1.3) The case discusses four types of funding risk facing LTCM:

### The case discusses specific ways in which LTCM manages each of these risks. Briefly discuss them.

LTCM managed the four types of funding risks in the following ways:

Collateral Haircuts:In market disruptions, collateral haircuts could increase, leading to unfavorable terms for LTCM in funding a spread trade.
To manage this, LTCM obtained 100% financing on a fully collateralized basis for most trades. They also conducted stress tests on haircuts across various asset classes to anticipate and manage potential risks.

Repo Maturity: In adverse situations where LTCM's credit risk increased, securing longer-term repos, which were favorable for their trades, could become challenging.
LTCM countered this by engaging in relatively long-maturity repos, contrary to the industry norm. While much of their repos were overnight, they typically used contracts with a maturity of 6-12 months, managing their aggregate repo maturity to mitigate risks.

Equity Redemption: LTCM faced redemption risk from investors, especially if in a convergence trade, the two securities diverged significantly before converging. This could force LTCM to finance their strategies during margin calls or unwind positions at unfavorable rates, leading to capital losses.
To manage this, LTCM restricted redemptions of equity year by year, with a strong policy where unredeemed money was re-locked, minimizing the risk related to equity funding due to high leverage.

Loan Access: During crises, accessing loans could become more difficult, potentially worsening the fund's performance.
LTCM mitigated this risk by negotiating a revolving loan that did not include a Material Adverse Change clause. This ensured that the availability of debt funding was less correlated with the fund's performance, providing more stability in crisis situations.

### 1.4) LTCM is largely in the business of selling liquidity and volatility. Describe how LTCM accounts for liquidity risk in their quantitative measurements.

LTCM required counterparties to maintain the collateral balance through a daily two-way mark to market process. This approach ensured that the cash flow coming in from the counterparties' mark to market would fund LTCM's outflow for the mark to market call on their offsetting position. This mechanism helped in managing liquidity risks by ensuring a steady flow of funds.

LTCM also estimated theoretical worst-case haircuts it could face in adverse market situations. By forecasting these worst-case liquidity scenarios, LTCM could better structure its financing, reducing the likelihood of having to liquidate positions solely due to adverse market events. This foresight in planning provided a buffer against unexpected liquidity squeezes.

To account for liquidity risk quantitatively, LTCM adjusted security correlations in their models. For short-term horizons, they assumed a positive correlation between all trade categories. This means that even if their net exposure to a strategy changed sides, they still assumed a positive correlation to the new net position. This approach helped in better understanding and managing the potential impacts of liquidity changes on their portfolio.

### 1.5) Is leverage risk currently a concern for LTCM?

As of the current situation, leverage risk is not an immediate concern for LTCM due to the absence of extreme market events. However, it remains a potential threat. The considerable size of their committed capital, coupled with limited opportunities to effectively utilize excess capital for enhancing returns, has led LTCM to contemplate returning some of the investments made. This step would effectively reduce the leverage, thereby mitigating the associated risks to some extent.

### 1.6) Many strategies of LTCM rely on converging spreads. LTCM feels that these are almost win/win situations because of the fact that if the spread converges, they make money. If it diverges, the trade becomes even more attractive, as convergence is still expected at a future date. <br> <br> What is the risk in these convergence trades?

LTCM's positions were vulnerable to liquidity risk. If market liquidity reduced or markets became segmented, the divergent spreads could persist for extended periods. This risk materialized for LTCM, where trades that eventually paid off did so only after the firm experienced significant financial distress.

LTCM believed it could exit these convergence trades if they became too unprofitable using stop-loss orders. However, a stop-loss order isn't equivalent to a put option. If the price moves discontinuously, bypassing the stop-loss level, the order becomes ineffective. This means the firm could not rely on stop-loss orders as a fail-safe exit strategy.

Besides,in situations where the market is paralyzed or illiquid, executing a stop-loss order could be challenging. Unlike a put option, which is not concerned with price impact, a stop-loss order is subject to it. This could result in LTCM having to sell securities at the worst possible timing, leading to unfavorable market outcomes.

Last, although LTCM expected the spreads to converge eventually, there was a risk that this convergence might not happen as anticipated or within a feasible timeframe. This was exemplified about a year after the case study period, when LTCM lost most of its value due to trades that did not converge as expected.

In [29]:
import pandas as pd
import numpy as np
import scipy.stats as stats
from scipy.stats import kurtosis, skew
from scipy.stats import norm
import seaborn as sns
import statsmodels.api as sm
from statsmodels.regression.rolling import RollingOLS
import warnings
warnings.filterwarnings("ignore")

%matplotlib inline


import matplotlib.pyplot as plt
plt.rcParams['figure.figsize']=[15, 6]
import matplotlib.cm as cm

In [30]:
def performance_summary(return_data):

    summary_stats = return_data.mean().to_frame('Mean').apply(lambda x: x*12)
    summary_stats['Volatility'] = return_data.std().apply(lambda x: x*np.sqrt(12))
    summary_stats['Sharpe Ratio'] = summary_stats['Mean']/summary_stats['Volatility']
    
    summary_stats['Skewness'] = return_data.skew()
    summary_stats['Excess Kurtosis'] = return_data.kurtosis()
    summary_stats['VaR (0.5)'] = return_data.quantile(.05, axis = 0)
    summary_stats['CVaR (0.5)'] = return_data[return_data <= return_data.quantile(.05, axis = 0)].mean()
    summary_stats['Min'] = return_data.min()
    summary_stats['Max'] = return_data.max()
    
    wealth_index = 1000*(1+return_data).cumprod()
    previous_peaks = wealth_index.cummax()
    drawdowns = (wealth_index - previous_peaks)/previous_peaks

    summary_stats['Max Drawdown'] = drawdowns.min()
    summary_stats['Peak'] = [previous_peaks[col][:drawdowns[col].idxmin()].idxmax() for col in previous_peaks.columns]
    summary_stats['Bottom'] = drawdowns.idxmin()
    
    recovery_date = []
    for col in wealth_index.columns:
        prev_max = previous_peaks[col][:drawdowns[col].idxmin()].max()
        recovery_wealth = pd.DataFrame([wealth_index[col][drawdowns[col].idxmin():]]).T
        recovery_date.append(recovery_wealth[recovery_wealth[col] >= prev_max].index.min())
    summary_stats['Recovery'] = recovery_date
    
    return summary_stats

In [31]:
def regression_based_performance(factor, fund_ret, rf, constant = True):

    if constant:
        X = sm.add_constant(factor)
    else:
        X = factor
    y = fund_ret

    model = sm.OLS(y, X, missing='drop').fit()

    if constant:
        beta = model.params[1:]
        alpha = round(float(model.params['const']), 6) * 12
    
    else:
        beta = model.params
    
    treynor_ratio = ((fund_ret - rf).mean()*12)/beta[0]
    tracking_error = (model.resid.std()*np.sqrt(12))

    if constant:
        information_ratio = model.params[0]*12/tracking_error
    r_squared = model.rsquared
    if constant:
        return (beta, treynor_ratio, information_ratio, alpha, r_squared, tracking_error, model.resid)
    else:
        return (beta, treynor_ratio, r_squared, tracking_error, model.resid)
    


In [32]:
ltcm_returns = pd.read_excel(r"ltcm_exhibits_data.xlsx", header = [2], sheet_name='Exhibit 2')
ltcm_returns = ltcm_returns[1:].rename(columns={'Unnamed: 0':'Date'})
ltcm_returns = ltcm_returns.set_index('Date')
ltcm_returns = ltcm_returns.loc[:,['Fund Capital ($billions)','Gross Monthly Performancea','Net Monthly Performanceb','Index of Net Performance']]
ltcm_returns = ltcm_returns[:53].rename(columns={'Gross Monthly Performancea':'Gross Monthly Performance','Net Monthly Performanceb':'Net Monthly Performance'})
ltcm_returns.index = pd.to_datetime(ltcm_returns.index) + pd.offsets.MonthEnd(0) 


In [33]:
gmo_total_ret = pd.read_excel(r"gmo_analysis_data.xlsx", sheet_name='returns (total)', index_col = 0)
spy_total_ret = gmo_total_ret[['SPY']]
spy_total_ret.index.name = 'Date'

rf = pd.read_excel(r"gmo_analysis_data.xlsx", sheet_name = 'risk-free rate', index_col=0)
rf.index.name = 'Date'

spy_excess_ret = (spy_total_ret['SPY'] - rf['US3M']).to_frame('SPY')

ltcm_returns['Net Excess Returns'] = ltcm_returns['Net Monthly Performance'] - rf['US3M']
ltcm_returns['Gross Excess Returns'] = ltcm_returns['Gross Monthly Performance'] - rf['US3M']

spy_er_ltcm = spy_excess_ret[ltcm_returns.index[0]:ltcm_returns.index[-1]]
spy_er_ltcm.tail()

SPY
Date                
1998-03-31  0.044458
1998-04-30  0.008624
1998-05-31 -0.024961
1998-06-30  0.038341
1998-07-31 -0.017764

## 2) LTCM Risk Decomposition

- <h4> On Canvas, find the data file, <code>ltcm exhibits data.xlsx</code>. Get the gross and net (total) returns of LTCM from "Exhibit 2".</h4> <br>
- <h4>  Get the returns on SPY as well as the risk-free rate from the file, <code>gmo analysis data</code>.</h4> 

### 2.1) Summary stats.
### 2.1.a) For both the gross and net series of LTCM excess returns, report the mean, volatility, and Sharpe ratios. (Annualize them.)

In [34]:
ltcm_summary = performance_summary(ltcm_returns.loc[:, ['Gross Excess Returns','Net Excess Returns']])
ltcm_summary.loc[:, ['Mean', 'Volatility', 'Sharpe Ratio']]

Mean  Volatility  Sharpe Ratio
Gross Excess Returns  0.242077    0.136232      1.776946
Net Excess Returns    0.155360    0.111765      1.390059

### 2.1.b) Report the skewness, kurtosis, and (historic) VaR(.05).

In [35]:
ltcm_summary.loc[:,['Mean', 'Volatility', 'Sharpe Ratio', 'Skewness','Excess Kurtosis','VaR (0.5)']]

Mean  Volatility  Sharpe Ratio  Skewness  \
Gross Excess Returns  0.242077    0.136232      1.776946 -0.287725   
Net Excess Returns    0.155360    0.111765      1.390059 -0.810239   

                      Excess Kurtosis  VaR (0.5)  
Gross Excess Returns         1.586625  -0.030445  
Net Excess Returns           2.926921  -0.026415

### 2.1.c) Comment on how these stats compare to SPY and other assets we have seen. How much do they differ between gross and net?

In [36]:
spy_er_summary = performance_summary(spy_er_ltcm)
spy_er_summary.loc[:, ['Mean','Volatility','Sharpe Ratio','Skewness','Excess Kurtosis','VaR (0.5)']]

Mean  Volatility  Sharpe Ratio  Skewness  Excess Kurtosis  VaR (0.5)
SPY  0.173823    0.112294      1.547923 -0.433516        -0.362022   -0.04636

LTCM displays a higher return compared to SPY, and remarkably, this is achieved with a volatility level very similar to that of SPY. This results in a slightly higher Sharpe ratio for LTCM (net of fees and other charges), indicating a more efficient return on risk taken compared to SPY.

When considering the excess net returns, LTCM underperforms SPY, despite having similar levels of volatility. This leads to a slightly lower Sharpe ratio for LTCM in this regard, suggesting less efficiency in generating excess returns compared to SPY.

LTCM's Net returns exhibit more negative skewness and a significantly fatter tail compared to SPY. This implies the presence of more substantial negative monthly returns during the sample period for LTCM.
However, the lower Value at Risk (VaR) for LTCM compared to SPY suggests that these heavy negative returns are less frequent. This indicates that while LTCM may experience more severe losses when they occur, such events are rarer compared to SPY.

### 2.2) Using the series of net LTCM excess returns, denoted $\tilde{r}^{LTCM}$, estimate the following regression: <br>

\begin{align}
\tilde{r}^{LTCM} = \alpha + \beta^{m}\tilde{r}^{m}_{t} +\epsilon_t
\end{align}

### 2.2.a) Report  $\alpha$ and $\beta^{m}$. Report the $R^{2}$ stat.

In [37]:
factor = spy_er_ltcm['SPY']
reg_sum = []
for rets in ['Gross Excess Returns','Net Excess Returns']:
    fund_ret = ltcm_returns.loc[:, [rets]]
    reg = regression_based_performance(factor, fund_ret, 0)
    beta_mkt = reg[0][0]
    treynor_ratio = reg[1][0]
    information_ratio = reg[2]
    alpha = reg[3]
    r_squared = reg[4]
    reg_sum.append(pd.DataFrame([[alpha, beta_mkt, r_squared, treynor_ratio, information_ratio]], columns = ['Alpha', 'Market Beta','R-Squared','Treynor Ratio','Information Ratio'], index=[rets]))

mkt_reg_sum = pd.concat(reg_sum)
mkt_reg_sum

Alpha  Market Beta  R-Squared  Treynor Ratio  \
Gross Excess Returns  0.210816     0.179845   0.021976       1.346034   
Net Excess Returns    0.131532     0.137114   0.018979       1.133077   

                      Information Ratio  
Gross Excess Returns           1.564766  
Net Excess Returns             1.188141

### 2.2.b) From this regression, does LTCM appear to be a "closet indexer"?

A low R-Squared value in the regression between LTCM's returns and SPY's returns indicates that there is a very low correlation between the two. 
The low beta value further supports the conclusion that LTCM's returns do not move in tandem with the market (SPY). A higher beta would indicate a stronger relationship with the market index, which is typical of a closet indexer.
LTCM's strategy primarily involves long-short trades, which inherently differ from the strategies employed by closet indexers. Closet indexers tend to have portfolios that closely resemble a market index but with slight modifications, while LTCM's approach of long-short trades signifies an active management strategy seeking to exploit specific market inefficiencies or opportunities, rather than tracking an index.

### 2.2.c) From the regression, does LTCM appear to deliver excess returns beyond the risk premium we expect from market exposure?


Based on the regression analysis and the noted significant annualized alpha of approximately 13.15%, it is evident that LTCM does indeed deliver excess returns beyond the risk premium expected from market exposure. This high alpha indicates that LTCM's returns are not solely attributable to the market's movements but also stem from the fund's specific strategies and management skills, resulting in returns that exceed those predicted by their market risk exposure.

### 2.3) Let's check for non-linear market exposure. Run the following regression on LTCM's net excess returns:<br>
    
\begin{align}
\tilde{r}^{LTCM} = \alpha + \beta_{1}\tilde{r}^{m}_{t} + \beta_{2}(\tilde{r}^{m}_{t})^{2} +\epsilon_t
\end{align}

### 2.3.a) Report $\beta_{1}\, \beta_{2}$, and the $R^{2}$ stat.

In [38]:
spy_er_ltcm['$SPY^{2}$'] = spy_er_ltcm['SPY']**2
factor_squared = spy_er_ltcm.loc[:, ['SPY', '$SPY^{2}$']]
quad_reg_sum = []
for rets in ['Gross Excess Returns','Net Excess Returns']:
    fund_ret = ltcm_returns.loc[:, [rets]]
    reg_squared = regression_based_performance(factor_squared, fund_ret, 0)
    beta_mkt = reg_squared[0][0]
    beta_mkt_squared = reg_squared[0][1]
    treynor_ratio = reg_squared[1][0]
    information_ratio = reg_squared[2]
    alpha = reg_squared[3]
    r_squared = reg_squared[4]
    quad_reg_sum.append(pd.DataFrame([[alpha, beta_mkt, beta_mkt_squared, r_squared, treynor_ratio, information_ratio]], columns = ['Alpha', 'Market Beta', '$SPY^{2}$ Beta', 'R-Squared','Treynor Ratio','Information Ratio'], index=[rets]))

mkt_quad_reg_sum = pd.concat(quad_reg_sum)
mkt_quad_reg_sum

Alpha  Market Beta  $SPY^{2}$ Beta  R-Squared  \
Gross Excess Returns  0.242388     0.219805       -2.586773   0.028458   
Net Excess Returns    0.155040     0.166878       -1.926746   0.024321   

                      Treynor Ratio  Information Ratio  
Gross Excess Returns       1.101330           1.805091  
Net Excess Returns         0.930984           1.404397

### 2.3.b) Does the quadratic market factor do much to increase the overall LTCM variation explained by the market?

The quadratic market factor in the regression analysis for LTCM does contribute to a slight improvement in the R-Squared value. However, despite the slight increase, the R-Squared value remains fairly low, which suggests that the relationship between LTCM's performance and market returns is not strongly pronounced. 

The increase in R-Squared could be partially attributed to multicollinearity, especially since the alpha in the regression also increases. This indicates that adding the quadratic market factor may lead to more unexplained returns rather than providing a clearer understanding of LTCM's return dynamics in relation to the market.

The significant negative beta coefficient for the squared SPY returns is a distinctive feature of this factor. Given that monthly returns are typically small, and their squared values even smaller, a larger magnitude of beta is required to fit these small returns appropriately. This aspect should be carefully considered in interpreting the regression results.

### 2.3.c) From the regression evidence, does LTCM's market exposure behave as if it is long market options or short market options?


Based on the regression evidence, particularly the negative beta to SPY^2 returns, it appears that LTCM's market exposure behaves as if it is short market options. 

The beta coefficient related to SPY can be interpreted as the delta of a market option. 

The beta coefficient related to SPY^2 can be interpreted as the gamma of market options. Gamma measures the rate of change in the delta with respect to changes in the underlying asset's price. 

Since LTCM exhibits a negative beta to SPY^2, it implies that they are short the positive gamma, or in other words, short the market options. 

### 2.3.d) Should we describe LTCM as being positively or negatively exposed to market volatility?

With a negative beta for SPY Squared, a large monthly return in the market would result in significant underperformance for LTCM. This indicates that LTCM's returns are adversely affected by large movements in the market.

Big deviations in the market lead to notable underperformance of LTCM, implying that their strategy is negatively impacted by high volatility.

### 2.4) Let's try to pinpoint the nature of LTCM's nonlinear exposure. Does it come more from exposure to up-markets or down-markets? Run the following regression on LTCM's net excess returns:<br>


\begin{align}
\tilde{r}^{LTCM} = \alpha + \beta\tilde{r}^{m}_{t} + \beta_{u}max(\tilde{r}^{m}_{t} - k_{1},0) + + \beta_{d}max(k_{2} - \tilde{r}^{m}_{t},0) +\epsilon_t
\end{align}

    
where $k_{1}$ = .03 and $k_{2}$ = -.03. (This is roughly one standard deviation of $\tilde{r}^{m}$.)

In [39]:
# generate call and put options
spy_er_ltcm['SPY_call'] = np.where(spy_er_ltcm['SPY'] > 0.03, spy_er_ltcm['SPY'] - 0.03, 0)
spy_er_ltcm['SPY_put'] = np.where(-0.03 - spy_er_ltcm['SPY'] >= 0, -0.03 - spy_er_ltcm['SPY'], 0)
factor = spy_er_ltcm.loc[:, ['SPY', 'SPY_call', 'SPY_put']]
call_reg_sum = []
# regression for call and put options
for rets in ['Gross Excess Returns','Net Excess Returns']:
    fund_ret = ltcm_returns.loc[:, [rets]]
    reg_call = regression_based_performance(factor, fund_ret, 0)
    beta_mkt = reg_call[0][0]
    beta_call = reg_call[0][1]
    beta_put = reg_call[0][2]
    treynor_ratio = reg_call[1][0]
    information_ratio = reg_call[2]
    alpha = reg_call[3]
    r_squared = reg_call[4]
    call_reg_sum.append(pd.DataFrame([[alpha, beta_mkt, beta_call, beta_put, r_squared, treynor_ratio, information_ratio]], columns = ['Alpha', 'Market Beta', 'Call Beta', 'Put Beta', 'R-Squared','Treynor Ratio','Information Ratio'], index=[rets]))

mkt_call_reg_sum = pd.concat(call_reg_sum)
mkt_call_reg_sum

Alpha  Market Beta  Call Beta  Put Beta  R-Squared  \
Gross Excess Returns  0.173796     0.608536   -1.03835  1.632452   0.063774   
Net Excess Returns    0.101232     0.466610   -0.78214  1.289575   0.055486   

                      Treynor Ratio  Information Ratio  
Gross Excess Returns       0.397803           1.318510  
Net Excess Returns         0.332956           0.931974

### 2.4.b) Is LTCM long or short the call-like factor? And the put-like factor?


Based on the regression betas, as the call beta is negative while the put beta is positive, LTCM seem to be short the call-like factor and long the put-like factor.

### 2.4.c) Which factor moves LTCM more, the call-like factor, or the put-like factor?


The analysis suggests that the put-like factor has a more significant impact on LTCM's returns compared to the call-like factor.

### 2.4.d) In the previous problem, you commented on whether LTCM is positively or negatively exposed to market volatility. Using this current regression, does this volatility exposure come more from being long the market's upside? Short the market's downside? Something else?

The combination of these two positions—short on the call-like factor and long on the put-like factor—contributes to LTCM's negative exposure to market volatility. This exposure is characterized by a potential underperformance during significant market upswings and a potential benefit or lesser impact during market downturns.

## 3) The FX Carry Trade

Find an Excel data file, `fx carry data.xlsx`. The file has two sets of data:
- Risk-free rates across 5 currencies, as measured by annualized 3-month LIBOR rates.
- Spot FX rates, as direct quotes to the USD. (Note that all currencies are quoted as USD per
the foreign currency.)


For use in the homework, note the following:
- For risk-free rate data, $\tilde{r}^{f,i}_{t,t+1}$, the rate is known and reported in the data at time t. <b> Namely,
any given date t in the data file is reporting both ${S}^{i}_{t}$ and $\tilde{r}^{f,i}_{t,t+1}$</b> <br>
- The theory says to use log risk-free rates. You have the risk-free rate in levels: use the following equation to convert them:<br>


\begin{align}
\tilde{r}^{f,i}_{t,t+1} = ln(1+ \tilde{r}^{f,i}_{t,t+1})
\end{align}


- The theory says to use log spot FX prices. You have the FX prices in levels, so directly take their logarithims:<br>

\begin{align}
{s}^{i}_{t} = ln({S}^{i}_{t})
\end{align}

In [40]:
risk_free_rates = pd.read_excel(r"fx_carry_data.xlsx", sheet_name = 'risk-free rates')
risk_free_rates.index = risk_free_rates['DATE']
risk_free_rates = risk_free_rates.drop(['DATE'], axis=1) # drop the date column

for col in risk_free_rates.columns:
    risk_free_rates['log_'+ col] = risk_free_rates[col].apply(lambda x: np.log(1+x)) # log return
risk_free_rates.head()

USD1M     GBP1M     EUR1M     CHF1M     JPY1M  log_USD1M  \
DATE                                                                      
1999-01-31  0.004116  0.004938  0.002598  0.000995  0.000335   0.004107   
1999-02-28  0.004135  0.004624  0.002600  0.001036  0.000232   0.004127   
1999-03-31  0.004114  0.004434  0.002492  0.000993  0.000143   0.004106   
1999-04-30  0.004085  0.004404  0.002141  0.000801  0.000099   0.004077   
1999-05-31  0.004120  0.004420  0.002141  0.000833  0.000075   0.004111   

            log_GBP1M  log_EUR1M  log_CHF1M  log_JPY1M  
DATE                                                    
1999-01-31   0.004926   0.002595   0.000995   0.000335  
1999-02-28   0.004614   0.002597   0.001036   0.000232  
1999-03-31   0.004424   0.002489   0.000992   0.000143  
1999-04-30   0.004394   0.002139   0.000801   0.000099  
1999-05-31   0.004411   0.002138   0.000833   0.000075

In [41]:
fx_rates = pd.read_excel(r"fx_carry_data.xlsx", sheet_name = 'fx rates')
fx_rates.index = fx_rates['DATE']
fx_rates = fx_rates.drop(['DATE'], axis=1) # drop the date column

for col in fx_rates.columns:
    fx_rates['log_'+ col] = fx_rates[col].apply(lambda x: np.log(x)) # log return
fx_rates.head()

USUK    USEU      USSZ      USJP  log_USUK  log_USEU  log_USSZ  \
DATE                                                                           
1999-01-31  1.6457  1.1371  0.705816  0.008621  0.498166  0.128481 -0.348401   
1999-02-28  1.6027  1.0995  0.689893  0.008425  0.471690  0.094856 -0.371219   
1999-03-31  1.6140  1.0808  0.676819  0.008444  0.478716  0.077702 -0.390351   
1999-04-30  1.6085  1.0564  0.655437  0.008373  0.475302  0.054867 -0.422453   
1999-05-31  1.6020  1.0422  0.654450  0.008273  0.471253  0.041334 -0.423960   

            log_USJP  
DATE                  
1999-01-31 -4.753590  
1999-02-28 -4.776599  
1999-03-31 -4.774322  
1999-04-30 -4.782730  
1999-05-31 -4.794798

### 3.1) The Static Carry Trade

#### 这里expected of excess log return = 0

Define the log return of holding the foreign currency using log values of the risk-free rate and
log values of the FX rates:<br>

\begin{align}
{r}^{i}_{t+1} = {s}^{i}_{t+1} - {s}^{i}_{t} + {r}^{f,i}_{t,t+1} 
\end{align}

Then the excess log return relative to USD, is expressed as

\begin{align}
\tilde{r}^{i}_{t+1} = {s}^{i}_{t+1} - {s}^{i}_{t} + {r}^{f,i}_{t,t+1} - {r}^{f,\$}_{t,t+1}
\end{align}

\begin{align}
\tilde{r}^{i}_{t+1} = {s}^{i}_{t+1} - {s}^{i}_{t} - ({r}^{f,\$}_{t,t+1} - {r}^{f,i}_{t,t+1})
\end{align}

For each foreign currency, i, calculate the excess log return series, $\tilde{r}_{t+1}$. Report the following
stats, (based on the excess log returns.) Annualize them.
- mean
- volatility
- sharpe ratio

In [42]:
#将exchange rate和risk free rate两个dataset的column name 合并
fx_spot_map = {'log_GBP1M':'log_USUK',
               'log_EUR1M':'log_USEU',
               'log_CHF1M':'log_USSZ',
               'log_JPY1M':'log_USJP'}

fx_hldg_lst = []
for k, v in fx_spot_map.items():
    fx_hldg_excess_ret = fx_rates[v] - fx_rates[v].shift(1) + risk_free_rates[k].shift(1) - risk_free_rates['log_USD1M'].shift(1)
    fx_hldg_summary = performance_summary(fx_hldg_excess_ret.to_frame().dropna())
    fx_hldg_summary.index = [k[4:7]]
    fx_hldg_summary.index.name = 'Currency Held'
    fx_hldg_lst.append(fx_hldg_summary)

fx_hldg_summary = pd.concat(fx_hldg_lst)
fx_hldg_summary.loc[:, ['Mean','Volatility','Sharpe Ratio', 'Min', 'Max']]


Mean  Volatility  Sharpe Ratio       Min       Max
Currency Held                                                        
GBP           -0.003502    0.086303     -0.040574 -0.094861  0.088332
EUR           -0.004351    0.094714     -0.045944 -0.103716  0.093510
CHF            0.004312    0.098757      0.043662 -0.118548  0.129996
JPY           -0.017415    0.091492     -0.190342 -0.085030  0.074347

#### 当sharpe ratio是负数说明不能invest in foreign currency 应该borrow from foreign currency to invest in USD

#### test whether the forward premium (excess return) is constant --> UIP regression test: beta (must be) = 1, alpha is positive --> although UIP does not hold, there is constant alpha
    - result: UIP does not hold and alpha (premium) varies through time
#### dynamic trading: if the premium gets larger, invest in foreign currency more, if gets smaller, invest less, if goes negative, short foreign currency --> 

#### dynamic trading of currency is not highly correlated with buy and hold of currency

### What differences do you see across currencies?

the CHF carry trade exhibits a positive and efficient return profile of 0.4% annualized return, while the GBP, EUR, and JPY carry trades show negative returns. The JPY carry trade, in particular, demonstrates the least favorable risk-return characteristics among the four currencies.

### 3.2) Implications for UIP:

### 3.2.a) Do any of these stats contradict the (log version) of Uncovered Interest Parity (UIP)?

The Uncovered Interest Parity (UIP) hypothesis, particularly its log version, posits that the mean return of currency positions should be zero, as the changes in the spot foreign exchange (FX) rate are expected to be completely offset by the differences in risk-free interest rates between two currencies. 

the Sharpe ratios for all currencies fall within the 95% confidence interval for statistical significance, it implies that the mean return statistics might not be significantly different from zero.
This means that, while the mean returns are not exactly zero, they might not be statistically significant enough to conclusively contradict UIP.

### 3.2.b) A long position in which foreign currency offered the best Sharpe ratio over the sample?

Based on the output above, a long position in CHF would have offered positive returns and a positive Sharpe ratio at 0.04.

### 3.2.c) Are there any foreign currencies for which a long position earned a negative excess return (in USD) over the sample?

GBP, EUR, and JPY earned negative excess returns in USD if taking long positions in those currencies. Especially for JPY, it had significant negative returns over the sample. 

### 3.3) Predicting FX

For each foreign currency, test whether interest-rate differentials can predict growth in the foreign-exchange rate.1 Do this by estimating the following forecasting regression:

\begin{align}
{s}^{i}_{t+1} - {s}^{i}_{t} = \alpha^{i} + \beta^{i}({r}^{f,\$}_{t,t+1} -  {r}^{f,i}_{t,t+1})+\epsilon_{t+1}
\end{align}

where  ${r}^{f,i}$ denotes the risk-free rate of currency i, and $s^{i}$ denotes the FX rate for currency i.<br>
Again, note that both ${r}^{f,\$}_{t,t+1}$ and $s_{t}$ are determined at time t.

### 3.3.a) Make a table with columns corresponding to a different currency regression. Report the regression estimates $\alpha^{i}$ and $\beta^{i}$ in the first two rows. Report the $R^{2}$ stat in the third row.

In [43]:
fx_regression_data_list = []

for k, v in fx_spot_map.items():
    fx_regression_data = pd.DataFrame({
        'Dependent Variable': fx_rates[v].diff(),
        'Independent Variable': risk_free_rates['log_USD1M'] - risk_free_rates[k]
    }).dropna()
    
    regression_result = regression_based_performance(fx_regression_data['Independent Variable'], fx_regression_data['Dependent Variable'], 0)
    
    alpha = regression_result[3]
    beta = regression_result[0][0]
    r_squared = regression_result[4]
    
    fx_regression_data_list.append({
        'Currency': k[4:7],
        'Alpha': alpha,
        'Beta': beta,
        'R-Squared': r_squared
    })

fx_regression_results = pd.DataFrame(fx_regression_data_list)
fx_regression_results.index = fx_regression_results['Currency']
fx_regression_results = fx_regression_results.drop(['Currency'], axis=1)
fx_regression_results.T


Currency        GBP       EUR       CHF      JPY
Alpha     -0.007596  0.008868  0.049584 -0.00114
Beta       0.110666 -1.633806 -2.066667  0.10538
R-Squared  0.000020  0.004398  0.006156  0.00004

#### beta is positive but less than 1: interest rate differential (r$ - r€) does predict exchange rate but just a little

#### beta is negative: added to future affect movement

#### R^2 is low means you can only make money when you trades a lot of times 

In [44]:
'''
fx_regression_results = pd.DataFrame(columns=['Currency', 'Alpha', 'Beta', 'R-Squared'])

for k,v in fx_spot_map.items():
    fx_regression_data = pd.DataFrame({
        'Dependent Variable': fx_rates[v].diff(),
        'Independent Variable': risk_free_rates['log_USD1M'] - risk_free_rates[k]
    }).dropna()
    
    regression_result = regression_based_performance(fx_regression_data['Independent Variable'], fx_regression_data['Dependent Variable'], 1)
    
    alpha = regression_result[3]
    beta = regression_result[0][0]
    r_squared = regression_result[4]
    
    # NOT APPEND HERE?????
    fx_regression_results = fx_regression_results.append({
        'Currency': k[4:7],
        'Alpha': alpha,
        'Beta': beta,
        'R-Squared': r_squared
    }, ignore_index=True)

fx_regression_results.index = fx_regression_results['Currency']
fx_regression_results = fx_regression_results.drop(['Currency'], axis=1)
fx_regression_results
'''

"\nfx_regression_results = pd.DataFrame(columns=['Currency', 'Alpha', 'Beta', 'R-Squared'])\n\nfor k,v in fx_spot_map.items():\n    fx_regression_data = pd.DataFrame({\n        'Dependent Variable': fx_rates[v].diff(),\n        'Independent Variable': risk_free_rates['log_USD1M'] - risk_free_rates[k]\n    }).dropna()\n    \n    regression_result = regression_based_performance(fx_regression_data['Independent Variable'], fx_regression_data['Dependent Variable'], 1)\n    \n    alpha = regression_result[3]\n    beta = regression_result[0][0]\n    r_squared = regression_result[4]\n    \n    # NOT APPEND HERE?????\n    fx_regression_results = fx_regression_results.append({\n        'Currency': k[4:7],\n        'Alpha': alpha,\n        'Beta': beta,\n        'R-Squared': r_squared\n    }, ignore_index=True)\n\nfx_regression_results.index = fx_regression_results['Currency']\nfx_regression_results = fx_regression_results.drop(['Currency'], axis=1)\nfx_regression_results\n"

### 3.3.b) Suppose the foreign risk-free rate increases relative to the US rate.

### i. For which foreign currencies would we predict a relative strengthening of the USD in the following period?

A positive beta in the regression analysis indicates that if the risk-free rate of a currency increases relative to the US rate, there will be a decrease in the FX rates (USD per foreign currency). This decrease translates to a relative strengthening of the USD against that foreign currency.

Since only the Japanese Yen (JPY) and the British Pound (GBP) exhibit a positive beta in the regression analysis, we can predict a relative strengthening of the USD against these two currencies in the following period, assuming their respective risk-free rates increase relative to the US rate.

Therefore, in scenarios where the risk-free rates in Japan and the United Kingdom increase relative to the US, the USD is expected to strengthen against the JPY and GBP. Conversely, a dollar would buy more Japanese Yen and British Pounds in this case.

### ii. For which currencies would we predict relative weakening of the USD in the following period?

Since the Euro (EUR) and the Swiss Franc (CHF) both have a negative beta in the regression analysis, we can predict that the USD would relatively weaken against these two currencies in the following period, assuming their respective risk-free rates increase compared to the US rate.

### iii. This FX predictability is strongest in the case of which foreign currency?

Given that CHF has the largest R^2 among all currencies, CHF shows the strongest FX predictability among the analyzed currencies, but the actual strength of this predictability might still be considered weak if the R^2 value is not sufficiently high. This suggests caution in relying heavily on these predictions for significant financial decisions.

### 3.4) The Dynamic Carry Trade

Use this to write $\mathbb{E}_{t}[\tilde{r}^{i}_{t+1}]$ as a function of the interest-rate differential as well as $\alpha$ and $\beta$ from this FX regression.<br>


\begin{align}
\mathbb{E}_{t}[{s}_{t+1} - {s}_{t}] = \alpha + \beta({r}^{f,\$}_{t,t+1} -  {r}^{f,i}_{t,t+1})
\end{align}

Then use the definition of excess (log) returns on FX:<br>

\begin{align}
\tilde{r}^{i}_{t+1} = {s}_{t+1} - {s}_{t} - ({r}^{f,\$}_{t,t+1} -  {r}^{f,i}_{t,t+1})
\end{align}


Rearranging, this implies the following forecast for excess log returns:<br>

\begin{align}
\mathbb{E}_[\tilde{r}^{i}_{t+1}] = \alpha + (\beta-1) ({r}^{f,\$}_{t,t+1} -  {r}^{f,i}_{t,t+1})
\end{align}



### 3.4.a) Use your regression estimates from Problem 3 along with the formula above to calculate the fraction of months for which the estimated FX risk premium positive. That is, for each i, calculate how often in the time-series we have <br>

\begin{align}
\mathbb{E}_{t}[\tilde{r}^{i}_{t+1}] > 0
\end{align}

In [45]:
fx_prem_lst = []
for k,v in fx_spot_map.items():
    fx_er_usd = (risk_free_rates['log_USD1M'].shift(1) - risk_free_rates[k].shift(1)).to_frame('ER_over_USD')
    expected_fx_premium = float(fx_regression_results.loc[[k[4:7]], 'Alpha'])/12 + (fx_er_usd.loc[:,['ER_over_USD']]  * float(fx_regression_results.loc[[k[4:7]], 'Beta'] - 1))
    expected_fx_premium = expected_fx_premium.rename(columns={'ER_over_USD':k[4:7]})
    positive_premium =  len(expected_fx_premium[expected_fx_premium[k[4:7]] > 0])
    fx_prem_lst.append(pd.DataFrame([[positive_premium,len(expected_fx_premium),positive_premium*100/len(expected_fx_premium)]],columns=['Months - Positive Premium','Total Months','Frequency(%)-Positive Premium'],index=[k[4:7]]))
fx_premium = pd.concat(fx_prem_lst)
fx_premium



Months - Positive Premium  Total Months  Frequency(%)-Positive Premium
GBP                         70           274                      25.547445
EUR                        140           274                      51.094891
CHF                        172           274                      62.773723
JPY                          1           274                       0.364964

### 3.4.b) Which currencies most consistently have a positive FX risk premium? And for which currencies does the FX risk premium most often go negative?

CHF and EUR are the currencies that most consistently offer a positive FX risk premium, making them more attractive from a risk-reward perspective. In contrast, JPY and GBP are more likely to have a negative FX risk premium, suggesting a less favorable risk-return trade-off for these currencies in the sample period.

### 3.4.c) Explain how we could use these conditional risk premia to improve the static carry trade returns calculated in Problem 1.

The analysis in 3.4.a) suggests that JPY returns tend to deviate from the expected value of zero, indicating a potential opportunity.
An effective strategy could involve shorting the JPY, meaning borrowing in JPY at the Japanese risk-free rate and investing those funds in an asset with a higher risk-free rate, such as the USD. 

---